design-decisions/2018-07-25 Attributes as patterns.md

```diff
@@ -0,0 +1,146 @@
+
+Similarly, we may want to use an `refersTo`
+attribute with a symbol as argument to, say,
+link an associativity axiom with the symbol.
+
+However, we can't tell what's the general structure of an attribute.
+
+Attribute syntax
+----------------
+
+We could encode each attribute as a string (name/key) or as a pair of
+strings (key-value).
+However, that does not seem a natural way of encoding
+the structure that they have, and each backend would need its own
+string-parsing library.
+Also, we may get some performance benefits since looking up a string key in a
+map is slower than looking up a constant key from a finite and (likely)
+very small set of such constants. However, note that the design of
+backend-specific data structures for attributes may remove the need for
+lookups.
+
+We could also encode attributes as Kore patterns which must be fully correct.
+This has some drawbacks:
+* We would need a central definition of all attributes that might be used by
+  backends, and all backends would need to ber aware of this definition.
+  This certainly has some benefits (e.g. one can find all the
+  attributes and their documentation in a single place).
+  As long as all attributes are generated by the frontend,
+  this may be feasible, but having such a central repository would mean
+  that independent backends cannot introduce their own attributes.
+* Mentioning symbols and sorts in attributes becomes harder and unnatural.
+  E.g. we wouldn't just be able to write `ac_axiom_for{}(plus{}())`, we would
+  need to write either `ac_axiom_for{}(\dv(symbol{}("plus"))` which means that
+  we partly give up on having structured data or
+  `ac_axiom_for{}(plus{}(x:Int{},y:Int{}))`
+  which would make it harder to manipulate these attributes.
+* The syntax may be hard to read.
+  This is not that important since K is meant to be readable, Kore is meant
+  to be machine-readable.
+
+*Decision*: We will encode each attribute as a Kore pattern that uses only
+application and string literals. This pattern will be an application
+of an object symbol to a list of parameters,
+the (name of the) symbol identifying the
+attribute and the list of parameters being the arguments of the attribute.
+Numbers will be encoded as string literals.
+Therefore the following may be valid attributes:
+```
+constructor{}()
+strict{}("1", "2")
+unit{}(epsilon{}())
+unit{}(epsilon{}(), zero{}())
+refersTo{}(plus{}())
+```
+We will *not* check that this attributes are well-formed in the usual sense,
+e.g. we will not check whether the `plus` symbol in the example above has
+two arguments or that the `strict` symbol arguments are object patterns or have
+the `#Pattern` sort.
+We *will* check that the top-level element is an application pattern since we
+must be able to find its name.
+We *may* check that the pattern uses only application and string literals.
+We *will* require a backend to provide a validation function for each attribute
+it can understand.
+
```

Comment:
It looks to me that the only mendatory checking happening is to check that attributes are symbol application patterns. IMO we could even drop that checking, too; but I'm fine with leaving it here.

Reply:
Also the last requirement seems a bit redundant, since the backend may provide a idle validation function that checks nothing. Having this requirement doesn't enforce anything to happen.